### PR TITLE
`Programming exercises`: Display hidden files when editing as instructor and viewing the repository

### PR DESCRIPTION
**Pull request by pzdr7**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

The code editor currently filters out files based on their name, e.g. files that start with `.` or files that have unsupported extensions. This is inconvenient when editing the exercise in the online code editor and viewing the repository, as parts of the repository may be hidden by these rules.

Furthermore, hidden files do not appear when viewing the commit history in the repository view.

### Description

This PR adds some new functionality to display these files:

- Server:
  - New filter for retrieving files and directories
  - When retrieving the repository, all files are now retrieved - except for the .git directory.
- Client:
  - New service that checks whether a file's contents are likely binary
  - Binary files are not loaded into the code editor
  - The code editor now displays hidden files
    - when an instructor edits the exercise (Edit in Editor)
    - when a user views the repository (Open Repository)
  - Hidden files are also visible when a user checks the commit history and selects a commit, except for binaries.

### Steps for Testing

Prerequisites:
- 1 Test Server with LocalCI
- 1 Instructor
- 1 Student
- 1 Programming Exercise with the Online Code Editor enabled. Use the default Java template

**Edit in editor**

1. Log in to Artemis as the instructor
2. Navigate to Course Management > Your course > Your programming exercise > Edit in Editor.
3. Edit the template repository: verify that files such as .gitignore, .gitattributes are visible and that you can make changes.
4. Verify that you cannot load the gradle-wrapper.jar into the editor (but that it is visible in the file browser) and that an appropriate placeholder is displayed if you try.

**Participating**

5. Log in as the student
6. Navigate to your course > Your programming exercise > Open code editor
7. Verify that the files .gitignore, .gitattributes, and the gradle-wrapper.jar are not visible.

**Repository view**

8. Go back to the programming exercise.
9. Click "Open Repository". In the file browser, the hidden files should be visible.
10. Click "Open Commit History" and pick the initial commit
11. Verify that .gitattributes and .gitignore are visible and that the gradle-wrapper.jar is not listed.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [code-editor-monaco.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6493/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/code-editor/monaco/code-editor-monaco.component.ts.html) | 91.46% | ✅ |
| [file-type.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6493/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/service/file-type.service.ts.html) | 100% | ✅ |

### Screenshots

image

image

## Summary by CodeRabbit

- **New Features**
  - Enhanced file and directory filtering in the Git service.
  - Enabled support for hidden files in various code editor components across the platform.
  - Introduced conditional rendering for file selection messages in Monaco editor component.
  - Added functionality to identify binary file content.

- **Bug Fixes**
  - Adjusted file visibility settings to allow hidden files in the repository view and code editor components.

- **Tests**
  - Added tests to ensure binary files are not loaded into the Monaco editor.
  - Implemented a test suite for verifying binary file content identification.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 1]
```diff
@@ -1161,6 +1161,21 @@ public void anonymizeStudentCommits(Repository repository, ProgrammingExercise p
...
+        @Override
+        public boolean accept(java.io.File file) {
+            return !GIT_DIRECTORY_NAME.equals(file.getName());
+        }
+
+        @Override
+        public boolean accept(java.io.File directory, String fileName) {
+            return !GIT_DIRECTORY_NAME.equals(directory.getName());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:1177]: We want to exclude exactly the files from that directory, which is why we do not use the `fileName` here.

**Review by Strohgelaender — APPROVED** [review]
code lgtm

**Comment by b-fein** [comment]
Did not look at the code, yet, but from the description it looks like this solves #4868. Nice!

**Comment by krusche** [comment]
> Did not look at the code, yet, but from the description it looks like this solves #4868. Nice!

For now, we show those files in the readonly code editor (which replaces the Bitbucket view) and in the instructor code editor (used with "Edit in Editor"). We still need to discuss whether we want to keep hiding certain files from students (such as the gradle wrapper) or not in the future. We are open for suggestions here.

The initial idea was to keep it simple for beginners and to avoid showing them all files. A compromise could be that those files are hidden by default, but we add a button "Show hidden files" (potentially with an icon) for experienced students so they can see all files. In any case, it won't be possible to edit binary files.

**Review by krusche — APPROVED** [review]
Code looks good 👍

**Review by coolchock — APPROVED** [review]
tested on ts2, works as described 👍

**Pull request merged**